### PR TITLE
SALTO-4599 - custom object field positions

### PR DESCRIPTION
**Pull request by idoser96**
_Replace me with a description of the changes in this PR_

---

_Additional context for reviewer_

---
_Release Notes_: 
_Replace me with a short sentence that describes the effect of this change on Salto users_

---
_User Notifications_: 
_Replace me with a short sentence that describes changes that will appear in NaCls and are not caused by user actions (e.g. a new annotation, field values that are converted to references, etc). Hidden changes should not be listed._

### CODE REVIEW

File: packages/zendesk-adapter/src/config.ts [hunk 1]
```diff
@@ -2562,11 +2562,7 @@ export const DEFAULT_TYPES: ZendeskApiConfig['types'] = {
   custom_object_fields: {
     request: {
       url: '/api/v2/custom_objects/{custom_object_key}/fields',
-      queryParams: {
-        ...DEFAULT_QUERY_PARAMS,
-        // Some fields are 'standard' and are not custom by the user, and won't be returned unless explicitly requested
-        include_standard_fields: 'true',
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/config.ts]: why did you decide to remove the standard fields?

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts [hunk 2]
```diff
@@ -372,7 +373,7 @@ const customObjectFieldsFilter: FilterCreator = ({ config, client }) => {
     },
     // Knowing if a value is a user depends on the custom_object_field attached to its condition's field
     // For that reason we need to specifically handle it here, using 'is_user_value' field that we added in onFetch
-    preDeploy: async (changes: Change<InstanceElement>[]) => {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:376]: why did you do this change?

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  CUSTOM_OBJECT_FIELD_TYPE_NAME,
+  CUSTOM_OBJECT_TYPE_NAME,
+  ZENDESK,
+} from '../../constants'
+
+const { RECORDS_PATH } = elementsUtils
+const log = logger(module)
+const { isDefined } = lowerDashValues
+
+export const ORDER_FIELD = `${CUSTOM_OBJECT_FIELD_TYPE_NAME}s`
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts]: can you move this to constants?

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      .filter(isInstanceElement)
+      .filter(instance => instance.elemID.typeName === CUSTOM_OBJECT_FIELD_TYPE_NAME)
+    const customObjectsByFullName = _.keyBy(
+      elements.filter(isInstanceElement).filter(instance => instance.elemID.typeName === CUSTOM_OBJECT_TYPE_NAME),
+      element => element.elemID.getFullName()
+    )
+
+    const customObjectFieldsByParentName = _.groupBy(
+      customObjectFields.filter(field => getParents(field).length === 1),
+      field => getParent(field).elemID.getFullName()
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts:68]:
```suggestion
      field => getParent(field)[0].elemID.getFullName()
```

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    )
+
+    if (!_.isEmpty(customObjectFieldsByParentName)) {
+      elements.push(customObjectFieldsOrderType)
+    }
+
+    Object.entries(customObjectFieldsByParentName).forEach(([parentName, fields]) => {
+      const parent = customObjectsByFullName[parentName]
+      const instanceName = `${parent.elemID.name}_fields_order`
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts:81]: parent can be undefined

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+      const orderInstance = new InstanceElement(
+        instanceName,
+        customObjectFieldsOrderType,
+        {
+          [ORDER_FIELD]: fields.map(field => new ReferenceExpression(field.elemID, field)),
+        },
+        [ZENDESK, RECORDS_PATH, CUSTOM_OBJECT_FIELD_ORDER_TYPE_NAME, instanceName],
+        {
+          [CORE_ANNOTATIONS.PARENT]: parent
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts]:
```suggestion
          [CORE_ANNOTATIONS.PARENT]: parent !== undefined
```

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        isInstanceChange(change)
+        && isAdditionOrModificationChange(change)
+        && getChangeData(change).elemID.typeName === CUSTOM_OBJECT_FIELD_ORDER_TYPE_NAME
+    )
+
+    const deployResultsPromises = relevantChanges
+      .filter(isInstanceChange) // used to type check
+      .map(async change => {
+        const customObjectFieldOrder = getChangeData(change)
+        const parentKey = getParent(customObjectFieldOrder).value.key
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts]: getParent can throw an error, either use try and catch or use getParents

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        if (parentKey === undefined) {
+          return {
+            change,
+            error: 'parent custom_object key is undefined',
+          }
+        }
+        const fieldsIds = customObjectFieldOrder.value[ORDER_FIELD]
+          .map((field: Value) => {
+            if (isResolvedReferenceExpression(field)) {
+              return field.value.value.id.toString()
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts:127]: we need to think of a situation where id is not defined
```suggestion
              return field.value.value.id?.toString()
```

File: packages/zendesk-adapter/test/filters/custom_objects/custom_object_fields_order.test.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    putSpy = jest.spyOn(client, 'put')
+    customObjectFieldsOrderFilter = filterCreator(createFilterCreatorParams({ client })) as FilterType
+    customObject1 = createCustomObject(1)
+    customObject2 = createCustomObject(2)
+    customObjectField1 = createCustomObjectField(1, customObject1)
+    customObjectField2 = createCustomObjectField(2, customObject1)
+    customObjectField3 = createCustomObjectField(3, customObject2)
+  })
+  describe('onFetch', () => {
+    it('should create order instance and order type', async () => {
```
Comment (edenhassid) [inline packages/zendesk-adapter/test/filters/custom_objects/custom_object_fields_order.test.ts:84]: can you add more tests, for cases where id is not defined, or when there is not parent

**Review by edenhassid — COMMENTED** [review]
looks good, added a few comments

File: packages/zendesk-adapter/src/config.ts (hunk 1 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/config.ts]: It is created automatically, and when i tried to create an element with these i received an error that they cant be created because they already exists

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts (hunk 2 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:376]: I think its a bad merge, fixing

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts (hunk 4 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts:68]: its getParent, not getParents, so it returns an instanceElement

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts (hunk 6 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields_order.ts]: parent exists for sure, changed it to just the parent, without the condition

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts (hunk 2 above)
Comment (edenhassid) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:376]: it wasn't fixed

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts [hunk 10]
```diff
@@ -408,7 +409,7 @@ const customObjectFieldsFilter: FilterCreator = ({ config, client }) => {
         }
       }
     },
-    onDeploy: async (changes: Change<InstanceElement>[]) => {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:411]: same here

File: packages/zendesk-adapter/src/filters/field_references.ts [hunk 11]
```diff
@@ -842,7 +842,15 @@ const commonFieldNameToTypeMappingDefs: ZendeskFieldReferenceDefinition[] = [
   // note: this overlaps with additional strategies, but because the first strategy
   // is chosen for serialization, it is safe
   {
-    src: { field: 'value', parentTypes: ['workspace__conditions__all'] },
+    src: {
+      field: 'value',
+      parentTypes: [
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/field_references.ts:847]: this is not related to this PR right?

File: packages/zendesk-adapter/src/filters/handle_template_expressions.ts [hunk 12]
```diff
@@ -166,7 +166,12 @@ const potentialTemplates: PotentialTemplateField[] = [
     containerValidator: (container: Values): boolean =>
       container.name === 'uri_templates',
   },
-  ...[ORG_FIELD_TYPE_NAME, USER_FIELD_TYPE_NAME, TICKET_FIELD_TYPE_NAME].flatMap(instanceType => [
+  ...[
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/handle_template_expressions.ts:169]: here too, this is related to this PR?

**Review by edenhassid — COMMENTED** [review]
added some comments

File: packages/zendesk-adapter/src/filters/field_references.ts (hunk 11 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/field_references.ts:847]: No, talked with Neta about it (its a fix on the way)

File: packages/zendesk-adapter/src/filters/handle_template_expressions.ts (hunk 12 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/handle_template_expressions.ts:169]: it is, ive added custom_object_field there

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts (hunk 2 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:376]: i got confused when i answered, i changed it to "onDeploy: async changes => {" because the previous one was incorrect
"onDeploy: async changes => {" is the correct one

File: packages/zendesk-adapter/e2e_test/adapter.test.ts [hunk 13]
```diff
@@ -1144,7 +1167,24 @@ describe('Zendesk adapter E2E', () => {
...
+          // toMatchObject does not work well with circular references and crashes
+          if ([CUSTOM_OBJECT_TYPE_NAME, CUSTOM_OBJECT_FIELD_TYPE_NAME].includes(instanceToAdd.elemID.typeName)) {
+            const instanceClone = (instance as InstanceElement).clone()
+            const instanceToAddClone = instanceToAdd.clone()
+            const fieldToHandle = instanceClone.elemID.typeName === CUSTOM_OBJECT_TYPE_NAME
+              ? `${CUSTOM_OBJECT_FIELD_TYPE_NAME}s`
+              : CUSTOM_FIELD_OPTIONS_FIELD_NAME
+
+            instanceClone.value[fieldToHandle] = (instanceClone.value[fieldToHandle] ?? [])
+              .map((ref: ReferenceExpression) => ref.value.elemID.getFullName())
```
Comment (edenhassid) [inline packages/zendesk-adapter/e2e_test/adapter.test.ts]: maybe check that it is a reference expression first? wdyt?

Comment (edenhassid) [inline packages/zendesk-adapter/e2e_test/adapter.test.ts]: if its a reference expression you don't need the value (same for the next lines as well)
```suggestion
              .map((ref: ReferenceExpression) => ref.elemID.getFullName())
```

**Review by edenhassid — COMMENTED** [review]
looks good! added some comments

File: packages/zendesk-adapter/e2e_test/adapter.test.ts (hunk 13 above)
Comment (idoser96) [reply packages/zendesk-adapter/e2e_test/adapter.test.ts]: was thinking about it, but because this is a test, the types are always the same (and if not it will crash on run)

**Review by edenhassid — APPROVED** [review]
LGTM!
